dask/highlevelgraph.py

```diff
@@ -523,6 +524,218 @@ def layer_info_dict(self):
+        produce keys corresponding to these collection indices.
+        """
+        raise NotImplementedError
+
+    def cull(self, keys, all_keys):
+        """Cull an AbstractLayer
+
+        The underlying graph will only include the necessary
+        tasks to produce the keys (indicies) included in `output_blocks`.
+        Therefore, "culling" the layer should only require us to reset
+        this parameter.
+        """
+        output_blocks = self._keys_to_indices(keys)
+        if output_blocks != set(self.output_blocks):
+            new_state = self.layer_state.copy()
+            new_state["output_blocks"] = output_blocks
+            new_state["annotations"] = self.annotations
+            culled_layer = self.reconstructor()(**new_state)
+        else:
+            culled_layer = self
+        return culled_layer, culled_layer.layer_dependencies(
+            keys, output_blocks=output_blocks
+        )
+
+    def is_materialized(self):
+        return hasattr(self, "_cached_dict")
+
+    @property
+    def _dict(self):
+        """Materialize full dict representation"""
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict
+        else:
+            dsk = self.construct_graph()
+            self._cached_dict = dsk
+        return self._cached_dict
+
+    def get_output_keys(self):
+        if hasattr(self, "_cached_output_keys"):
+            return self._cached_output_keys
+        else:
+            if self.output_blocks is None:
+                output_keys = self.keys()  # Materializes the graph!
+            else:
+                output_keys = {(self.name, block) for block in self.output_blocks}
+            self._cached_output_keys = output_keys
+        return self._cached_output_keys
+
+    def _keys_to_indices(self, keys):
+        """Convert keys to output chunk/partition indices
+
+        Currently works for keys of the form: `(name, index)`.
+        """
+        parts = set()
+        for key in keys:
+            try:
+                _name, _part = key
+            except ValueError:
+                continue
+            if _name != self.name:
+                continue
+            parts.add(_part)
+        return parts
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return iter(self._dict)
+
+    def __dask_distributed_pack__(self, all_hlg_keys, *args, **kwargs):
+        # from distributed.protocol.serialize import ToPickle
+        import cloudpickle
+
+        # Save "pre-stringified" key dependencies
+        # TODO: Is there a better way to do this?
+        state = self.layer_state.copy()
+        state["layer_dependencies"] = self.layer_dependencies(all_hlg_keys)
+
+        # TODO: Use ToPickle - The current implementation
+        # sometimes fails.
+        # return ToPickle(state)
+        return cloudpickle.dumps(state)
```

Comment:
@madsbk - I have been using `cloudpickle` while developing/testing `AbstractLayer`, because `ToPickle` sometimes fails here.

For example, when running `dask/tests/test_distributed.py::test_fused_blockwise_dataframe_merge[False]`:

```
Traceback (most recent call last):
  File "/home/rzamora/workspace/cudf-22.02/distributed/distributed/batched.py", line 94, in _background_send
    nbytes = yield self.comm.write(
  File "/home/rzamora/miniconda/miniconda3/envs/cudf_2204/lib/python3.8/site-packages/tornado/gen.py", line 762, in run
    value = future.result()
  File "/home/rzamora/workspace/cudf-22.02/distributed/distributed/comm/tcp.py", line 249, in write
    frames = await to_frames(
  File "/home/rzamora/workspace/cudf-22.02/distributed/distributed/comm/utils.py", line 50, in to_frames
    return _to_frames()
  File "/home/rzamora/workspace/cudf-22.02/distributed/distributed/comm/utils.py", line 33, in _to_frames
    return list(protocol.dumps(msg, **kwargs))
  File "/home/rzamora/workspace/cudf-22.02/distributed/distributed/protocol/core.py", line 102, in dumps
    frames[0] = msgpack.dumps(msg, default=_encode_default, use_bin_type=True)
  File "/home/rzamora/miniconda/miniconda3/envs/cudf_2204/lib/python3.8/site-packages/msgpack/__init__.py", line 35, in packb
    return Packer(**kwargs).pack(o)
  File "msgpack/_packer.pyx", line 294, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 300, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 297, in msgpack._cmsgpack.Packer.pack
  File "msgpack/_packer.pyx", line 264, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 231, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 231, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 264, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 231, in msgpack._cmsgpack.Packer._pack
  File "msgpack/_packer.pyx", line 285, in msgpack._cmsgpack.Packer._pack
  File "/home/rzamora/workspace/cudf-22.02/distributed/distributed/protocol/core.py", line 97, in _encode_default
    frames.extend(create_pickle_sub_frames(obj))
  File "/home/rzamora/workspace/cudf-22.02/distributed/distributed/protocol/core.py", line 81, in create_pickle_sub_frames
    _inplace_compress_frames(sub_header, sub_frames)
  File "/home/rzamora/workspace/cudf-22.02/distributed/distributed/protocol/core.py", line 45, in _inplace_compress_frames
    compression[i], frames[i] = maybe_compress(
  File "/home/rzamora/workspace/cudf-22.02/distributed/distributed/protocol/compression.py", line 198, in maybe_compress
    if len(payload) < min_size:
TypeError: object of type 'pickle.PickleBuffer' has no len()
```

With `cloudpickle`, everything in `test_distributed.py` passes for me locally.

Reply:
Should work now

Reply:
>Should work now

My environment may be messed up, but I am still seeing the same error